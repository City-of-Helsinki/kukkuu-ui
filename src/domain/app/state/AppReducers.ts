import { combineReducers } from '@reduxjs/toolkit';

import authenticationReducers from '../../auth/state/AuthenticationReducers';
import profileReducer from '../../profile/state/ProfileReducers';
import registrationReducer from '../../registration/state/RegistrationReducers';
import uiReducers from './ui/UIReducers';
import EventReducers from '../../event/state/EventReducers';

export default combineReducers({
  authentication: authenticationReducers,
  event: EventReducers,
  registration: registrationReducer,
  profile: profileReducer,
  ui: uiReducers,
});
