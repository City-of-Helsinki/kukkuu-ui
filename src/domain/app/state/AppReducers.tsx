import { combineReducers } from 'redux-starter-kit';

import registrationReducer from '../../registration/state/RegistrationReducers';
export default combineReducers({ registration: registrationReducer });
