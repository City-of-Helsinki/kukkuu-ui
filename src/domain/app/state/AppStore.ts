import { configureStore, getDefaultMiddleware, Store } from 'redux-starter-kit';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import rootReducer from './AppReducers';

const persistConfig = {
  key: 'root',
  storage,
  whitelist: ['registration'],
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store: Store = configureStore({
  devTools: true,
  middleware: getDefaultMiddleware({
    serializableCheck: false,
  }),
  reducer: persistedReducer,
});

const persistor = persistStore(store);

export { persistor, store };
