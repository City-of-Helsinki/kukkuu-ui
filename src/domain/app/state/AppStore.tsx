import { createStore } from 'redux';
import { devToolsEnhancer } from 'redux-devtools-extension';
import { loadUser } from 'redux-oidc';

import userManager from '../../auth/userManager';
import rootReducer from './AppReducers';
const store = createStore(rootReducer, devToolsEnhancer({}));

loadUser(store, userManager);

export default store;
