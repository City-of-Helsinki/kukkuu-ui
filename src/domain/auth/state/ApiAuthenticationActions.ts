import { createAction } from 'redux-starter-kit';

import { API_AUTHENTICATION_ACTIONS } from '../constants/ApiAuthenticationActionConstants';

const startFetchingToken = createAction(
  API_AUTHENTICATION_ACTIONS.START_FETCHING_TOKEN
);

const fetchTokenSuccess = createAction(
  API_AUTHENTICATION_ACTIONS.FETCH_TOKEN_SUCCESS
);

const fetchTokenError = createAction(
  API_AUTHENTICATION_ACTIONS.FETCH_TOKEN_ERROR
);

export { startFetchingToken, fetchTokenSuccess, fetchTokenError };
