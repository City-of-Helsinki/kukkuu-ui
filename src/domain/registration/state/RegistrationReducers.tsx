import { createReducer } from 'redux-starter-kit';

import { RegistrationData } from '../types/RegistrationTypes';
import { REGISTRATION_ACTIONS } from '../constants/RegistrationActionConstants';

export const defaultRegistrationData: RegistrationData = {
  formValues: {
    child: {
      birthday: '',
      firstName: '',
      homeCity: '',
      lastName: '',
    },
    guardian: {
      firstName: '',
      lastName: '',
      phoneNumber: '',
      relationship: '',
    },
    verifyInformation: false,
  },
};

export default createReducer(defaultRegistrationData, {
  [REGISTRATION_ACTIONS.SET_FORM_VALUES]: (state, action) =>
    Object.assign({}, state, { formValues: action.payload }),
  [REGISTRATION_ACTIONS.RESET_FORM_VALUES]: state =>
    Object.assign({}, state, {
      formValues: defaultRegistrationData.formValues,
    }),
});
