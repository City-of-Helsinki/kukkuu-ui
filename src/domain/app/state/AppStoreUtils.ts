import { persistor, store } from './AppStore';
import { resetBackendAuthentication } from '../../auth/state/BackendAuthenticationActions';
import client from '../../api/client';
import { clearProfile } from '../../profile/state/ProfileActions';
import { resetFormValues } from '../../registration/state/RegistrationActions';

/**
 * Flush entire Redux data store
 * as well as persisted data
 * Apollo cache
 */
export const flushData = () => {
  // Clear backend auth data
  store.dispatch(resetBackendAuthentication());
  // Clear user form data
  store.dispatch(resetFormValues());
  // Clear profile (fetched from API)
  store.dispatch(clearProfile());
  // Flush data in redux store and localStorage
  persistor.flush();
  // Clear Apollo cache
  client.clearStore();
};
