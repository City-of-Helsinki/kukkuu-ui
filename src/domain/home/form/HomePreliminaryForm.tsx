import React, { FunctionComponent } from 'react';
import { Formik, Field, FieldArray, FormikErrors } from 'formik';
import { connect } from 'react-redux';
import { useTranslation } from 'react-i18next';
import { useHistory } from 'react-router-dom';

import { loginTunnistamo } from '../../auth/authenticate';
import styles from './homePreliminaryForm.module.scss';
import Button from '../../../common/components/button/Button';
import InputField from '../../../common/components/form/fields/input/InputField';
import {
  validateRequire,
  validateBirthday,
} from '../../../common/components/form/validationUtils';
import BirthdayFormField from './partial/BirthdayFormField';
import { setFormValues } from '../../registration/state/RegistrationActions';
import { RegistrationFormValues } from '../../registration/types/RegistrationTypes';
import { defaultRegistrationData } from '../../registration/state/RegistrationReducers';
import { StoreState } from '../../app/types/AppTypes';
import { isAuthenticatedSelector } from '../../auth/state/AuthenticationSelectors';
import { HomeFormValues } from './types/HomeFormTypes';
import { convertFormValues } from './HomePreliminaryFormUtils';

interface Props {
  isAuthenticated: boolean;
  setFormValues: (values: RegistrationFormValues) => void;
  stateFormValues: RegistrationFormValues;
}

const HomePreliminaryForm: FunctionComponent<Props> = ({
  setFormValues,
  stateFormValues,
  isAuthenticated,
}) => {
  const { t } = useTranslation();
  const history = useHistory();

  const handleSubmit = (values: HomeFormValues) => {
    const defaultFormValues = defaultRegistrationData.formValues;
    const payload = Object.assign({}, defaultFormValues, {
      child: {
        birthday: `${values.child.birthday.day}.${values.child.birthday.month}.${values.child.birthday.year}`,
        homeCity: values.child.homeCity,
      },
      verifyInformation: values.verifyInformation,
    });

    setFormValues(payload);
    if (isAuthenticated) history.push('/registration/form');
    else loginTunnistamo(`/registration/form`);
  };

  const validate = (values: HomeFormValues) => {
    const {
      child: {
        birthday: { day, month, year },
      },
    } = values;
    const errors: FormikErrors<HomeFormValues> = {};

    if (day && month && year) {
      errors.childBirthday = validateBirthday(`${day}.${month}.${year}`);

      if (!errors.childBirthday) {
        // Delete the property manually so form will be valid when this is undefined.
        delete errors.childBirthday;
      }
    }
    return errors;
  };
  return (
    <div className={styles.homeForm}>
      <Formik
        initialValues={convertFormValues(stateFormValues)}
        onSubmit={handleSubmit}
        validate={validate}
      >
        {({ values, handleChange, handleSubmit, isSubmitting, isValid }) => (
          <form onSubmit={handleSubmit}>
            <div className={styles.inputWrapper}>
              <FieldArray
                name="child.birthday"
                render={props => <BirthdayFormField {...props} />}
              />

              <Field
                className={styles.childHomeCity}
                type="text"
                name="child.homeCity"
                label={t('homePage.preliminaryForm.childHomeCity.input.label')}
                onChange={handleChange}
                value={values.child.homeCity}
                component={InputField}
                placeholder={t(
                  'homePage.preliminaryForm.childHomeCity.input.placeholder'
                )}
              />
            </div>

            <Field
              className={styles.verifyInformationCheckbox}
              type="checkbox"
              label={t(
                'homePage.preliminaryForm.verifyInformation.checkbox.label'
              )}
              name="verifyInformation"
              onChange={handleChange}
              value={values.verifyInformation}
              checked={values.verifyInformation}
              component={InputField}
              validate={(value: boolean) =>
                validateRequire(
                  value,
                  'homePage.preliminaryForm.verifyInformation.checkbox.required.label'
                )
              }
            />

            <Button
              type="submit"
              className={styles.submitButton}
              disabled={isSubmitting || !isValid}
            >
              {t('homePage.hero.buttonText')}
            </Button>
          </form>
        )}
      </Formik>
    </div>
  );
};

const actions = {
  setFormValues,
};

const mapStateToProps = (state: StoreState) => ({
  isAuthenticated: isAuthenticatedSelector(state),
  stateFormValues: state.registration.formValues,
});

export const UnconnectedHomePreliminaryForm = HomePreliminaryForm;

export default connect(
  mapStateToProps,
  actions
)(HomePreliminaryForm);
