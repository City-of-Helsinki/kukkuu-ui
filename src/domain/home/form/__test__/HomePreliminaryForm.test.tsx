import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { UnconnectedHomePreliminaryForm } from '../HomePreliminaryForm';
import { defaultRegistrationData } from '../../../registration/state/RegistrationReducers';

it('renders snapshot correctly', () => {
  const element = shallow(
    <UnconnectedHomePreliminaryForm
      isAuthenticated={true}
      stateFormValues={defaultRegistrationData.formValues}
      setFormValues={jest.fn()}
    />
  );
  expect(toJson(element)).toMatchSnapshot();
});
