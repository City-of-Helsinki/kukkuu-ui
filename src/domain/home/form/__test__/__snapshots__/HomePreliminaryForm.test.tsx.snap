// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot correctly 1`] = `
<div
  className="homeForm"
>
  <Formik
    initialErrors={
      Object {
        "childBirthdate": "validation.date.invalidFormat",
      }
    }
    onSubmit={[Function]}
    validate={[Function]}
  >
    <Component />
  </Formik>
</div>
`;
